### ISSUE TITLE
create dataset _index - already exists! error

### ISSUE THREAD

**Issue by kvshams** [mojaveazure/seurat-disk#44]
I am trying to convert a H5Seurat object to h5 and getting an error as follows 
```
SaveH5Seurat(WT_KO, filename = "WT_KO_v5_RNAAssay.h5Seurat")
Convert("WT_KO_v5_RNAAssay.h5Seurat", dest = "h5ad", overwrite = TRUE)
```
```
Validating h5Seurat file

Adding scale.data from RNA as X

Adding data from RNA as raw

Transfering meta.data to obs

Error in dfile[["obs"]]$create_dataset(name = rownames, robj = Cells(x = source), : Can't create dataset _index - already exists!
Traceback:

1. Convert("WT_KO_v5_RNAAssay.h5Seurat", dest = "h5ad", overwrite = TRUE)
2. Convert.character("WT_KO_v5_RNAAssay.h5Seurat", dest = "h5ad", 
 .     overwrite = TRUE)
3. Convert(source = hfile, dest = dest, assay = assay, overwrite = overwrite, 
 .     verbose = verbose, ...)
4. Convert.h5Seurat(source = hfile, dest = dest, assay = assay, 
 .     overwrite = overwrite, verbose = verbose, ...)
5. H5SeuratToH5AD(source = source, dest = dest, assay = assay, overwrite = overwrite, 
 .     verbose = verbose)
6. dfile[["obs"]]$create_dataset(name = rownames, robj = Cells(x = source), 
 .     dtype = GuessDType(x = Cells(x = source)))
7. stop(paste("Can't create dataset", name, "- already exists!"))
```

```
R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.1 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] SeuratDisk_0.0.0.9018 SeuratObject_4.0.0    Seurat_4.0.0         

loaded via a namespace (and not attached):
  [1] Rtsne_0.15           colorspace_2.0-0     deldir_0.2-9        
  [4] ellipsis_0.3.1       ggridges_0.5.3       IRdisplay_0.7.0     
  [7] base64enc_0.1-3      spatstat.data_1.7-0  leiden_0.3.7        
 [10] listenv_0.8.0        farver_2.0.3         ggrepel_0.9.1       
 [13] bit64_4.0.5          codetools_0.2-18     splines_4.0.3       
 [16] polyclip_1.10-0      IRkernel_1.1.1.9000  jsonlite_1.7.2      
 [19] ica_1.0-2            cluster_2.1.0        png_0.1-7           
 [22] uwot_0.1.10          shiny_1.6.0          sctransform_0.3.2   
 [25] compiler_4.0.3       httr_1.4.2           assertthat_0.2.1    
 [28] Matrix_1.3-2         fastmap_1.1.0        lazyeval_0.2.2      
 [31] cli_2.3.0            later_1.1.0.1        htmltools_0.5.1.1   
 [34] tools_4.0.3          igraph_1.2.6         gtable_0.3.0        
 [37] glue_1.4.2           RANN_2.6.1           reshape2_1.4.4      
 [40] dplyr_1.0.4          Rcpp_1.0.6           spatstat_1.64-1     
 [43] scattermore_0.7      vctrs_0.3.6          nlme_3.1-151        
 [46] lmtest_0.9-38        stringr_1.4.0        globals_0.14.0      
 [49] mime_0.9             miniUI_0.1.1.1       lifecycle_0.2.0     
 [52] irlba_2.3.3          goftest_1.2-2        future_1.21.0       
 [55] MASS_7.3-53          zoo_1.8-8            scales_1.1.1        
 [58] promises_1.1.1       spatstat.utils_2.0-0 parallel_4.0.3      
 [61] RColorBrewer_1.1-2   reticulate_1.18      pbapply_1.4-3       
 [64] gridExtra_2.3        ggplot2_3.3.3        rpart_4.1-15        
 [67] stringi_1.5.3        repr_1.1.0           rlang_0.4.10        
 [70] pkgconfig_2.0.3      matrixStats_0.58.0   evaluate_0.14       
 [73] lattice_0.20-41      ROCR_1.0-11          purrr_0.3.4         
 [76] tensor_1.5           patchwork_1.1.1      htmlwidgets_1.5.3   
 [79] labeling_0.4.2       cowplot_1.1.1        bit_4.0.4           
 [82] tidyselect_1.1.0     parallelly_1.23.0    RcppAnnoy_0.0.18    
 [85] plyr_1.8.6           magrittr_2.0.1       R6_2.5.0            
 [88] generics_0.1.0       pbdZMQ_0.3-4         DBI_1.1.0           
 [91] pillar_1.4.7         withr_2.4.1          mgcv_1.8-33         
 [94] fitdistrplus_1.1-3   survival_3.2-7       abind_1.4-5         
 [97] tibble_3.0.6         future.apply_1.7.0   crayon_1.4.0        
[100] hdf5r_1.3.3          uuid_0.1-4           KernSmooth_2.23-18  
[103] plotly_4.9.3         grid_4.0.3           data.table_1.13.6   
[106] digest_0.6.27        xtable_1.8-4         tidyr_1.1.2         
[109] httpuv_1.5.5         munsell_0.5.0        viridisLite_0.3.0   
```

**Comment by OliverBancroft** [comment]
Same problems

**Comment by ChenXY-0919** [comment]
I have same problem.  I tried to solve the problem as following:
Just remove the block from line 1147 to 1151 of file "R/Convert.R" and rebuild the packages.
`  dfile[['obs']]$create_dataset( 
    name = rownames, 
    robj = Cells(x = source), 
    dtype = GuessDType(x = Cells(x = source)) 
  )`

**Comment by kvshams** [comment]
> I have same problem. I tried to solve the problem as following:
> Just remove the block from line 1147 to 1151 of file "R/Convert.R" and rebuild the packages.
> ` dfile[['obs']]$create_dataset( name = rownames, robj = Cells(x = source), dtype = GuessDType(x = Cells(x = source)) )`

Thanks @ChenXY-0919.. That worked, don't know the consequence though!

**Comment by joseah** [comment]
I'm running into the same issue. 

This problem seems to happen from the most recent commit [003cdf0](https://github.com/mojaveazure/seurat-disk/commit/003cdf0221e4ecd7a87e09e92e8f364ca782c2b6) on version `0.0.0.9018`, probably in the `WriteH5Group`. Downgrading to the previous commit seems to work `devtools::install_github("mojaveazure/seurat-disk@003cdf0")`.

**Comment by amasden** [comment]
Same problem here in Convert step showing:

Validating h5Seurat file
Adding scale.data from RNA as X
Transfering meta.features to var
Error in dfile[["var"]]$create_dataset(name = rownames, robj = assay.group[["features"]][x.features],  : 
  Can't create dataset _index - already exists!

**Comment by c5creative** [comment]
Ditto here, same:
```R
> Convert("filename.h5Seurat", dest = "h5ad")
Validating h5Seurat file
Adding scale.data from RNA as X
Transfering meta.features to var
Error in dfile[["var"]]$create_dataset(name = rownames, robj = assay.group[["features"]][x.features],  : 
  Can't create dataset _index - already exists!
```

**Comment by bapoorva** [comment]
I'm having the same issue. Downgrading didnt help.

**Comment by cornhundred** [comment]
I'm having the same issue and I'm working in the docker container

```
> Convert('docker_data/multi.h5seurat', dest='h5ad')
Validating h5Seurat file
Adding data from SCT as X
Adding counts from SCT as raw
Transfering meta.data to obs
Error in dfile[["obs"]]$create_dataset(name = rownames, robj = Cells(x = source),  : 
  Can't create dataset _index - already exists!
```

**Comment by SabrinaITNL** [comment]
Hi, 
Any updates on this error?
I still get it while trying to convert h5Seurat to h5ad.
Thanks!
Sabrina

**Comment by kvshams** [comment]
@SabrinaITNL you may try to install this fork `devtools::install_github("kvshams/seurat-disk")` . I simply modified the script  as discussed by @ChenXY-0919 and that worked perfectly fine. The only issue I faced is, many of the metadata columns lost its identity as charecters/factor and it created a mess in obs slot for some reason... That would easily could fix my exporting metadata separately as a table from `seurat object` and import it using `pandas` when we process the data in `python`... Hope this helps!

**Comment by akhst7** [comment]
@kvshams, I installed your fix of Covert in SeuratDisk, 
```
install_github("kvshams/seurat-disk")
Using github PAT from envvar GITHUB_PAT
Downloading GitHub repo kvshams/seurat-disk@HEAD

✓  checking for file ‘/private/var/folders/1v/3dyhjv_j0dvb5tn00xdkclbw0000gn/T/RtmpCihyQ7/remotes1b6d376e3319/kvshams-seurat-disk-e464736/DESCRIPTION’ (334ms)
─  preparing ‘SeuratDisk’:
✓  checking DESCRIPTION meta-information ...
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
─  building ‘SeuratDisk_0.0.0.9018.tar.gz’
```

but I am still getting the same error, 

```
Convert("pbmc10k.h5Seurat", dest = "h5ad" )
Validating h5Seurat file
Adding scale.data from RNA as X
Transfering meta.features to var
Error in dfile[["var"]]$create_dataset(name = rownames, robj = assay.group[["features"]][x.features],  : 
  Can't create dataset _index - already exists!
```

can you ckeck and make sure your repo has the corrected version ? 

Thanks.

**Comment by uberholzer** [comment]
Same issue here, @kvshams fix did not solve it

**Comment by jychien** [comment]
I tried reverting to SeuratDisk_0.0.0.9015: 
devtools::install_github("mojaveazure/seurat-disk@dae280a")

**Comment by akhst7** [comment]
@jychien,  the older version works but there is an new pull by sheetagirl that works as well.

**Comment by XinYu-pumch** [comment]
Got similar error with @c5creative. 
But when I reinstalled SeruatDisk by running the code @jychien provided (remotes::install_github("mojaveazure/seurat-disk@dae280a")), the project worked well.   

And remeber to delete the dataframe in your Seurat object@meta.data, if there is a dataframe.

**Comment by khushboojindal** [comment]
I got the similar issue. I got it fixed. Now, It's working for me.

Try this one :

devtools::install_github("mojaveazure/seurat-disk@63633d2")    (Install the older version of Seurat Disk i.e. Version: 0.0.0.9017, previously I was using  0.0.0.9018 that didn't work for me)
 
Make sure you have recent version of HDF5 package (Version: 1.3.3).  (Install newer version using devtools::install_github("hhoeflin/hdf5r"), previously I was using Version: 1.3.2 that didn't work for me)
